### PR TITLE
[New scheduler] Cluster Resource Scheduler dynamic resources (for placement groups) 

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This implements dynamic resources in the cluster resource scheduler which is necessary for creating the placement group resources. In order to do so it fixes a few latent bugs. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/scheduling/cluster_resource_data.cc [hunk 1]
```diff
@@ -83,7 +83,8 @@ TaskRequest ResourceMapToTaskRequest(
     } else if (resource.first == ray::kMemory_ResourceLabel) {
       task_request.predefined_resources[MEM].demand = resource.second;
     } else {
-      task_request.custom_resources[i].id = string_to_int_map.Insert(resource.first);
+      string_to_int_map.Insert(resource.first);
+      task_request.custom_resources[i].id = string_to_int_map.Get(resource.first);
```
Comment (wuisawesome) [inline src/ray/raylet/scheduling/cluster_resource_data.cc:87]: This was a latent bug. `string_to_int_map.Insert` only returns the ID if it already existed. Therefore, if this `Insert` call is the first call, everything breaks :(

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 2]
```diff
@@ -367,7 +395,8 @@ void ClusterResourceScheduler::UpdateResourceCapacity(const std::string &node_id
       it->second.predefined_resources[idx].total = 0;
     }
   } else {
-    int64_t resource_id = string_to_int_map_.Insert(resource_name);
+    string_to_int_map_.Insert(resource_name);
```
Comment (wuisawesome) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:398]: Same as the above bug.

**Comment by ericl** [comment]
Does this mean we can enable the dynamic set resource api tests?

**Comment by wuisawesome** [comment]
No, this is only the minimal implementation necessary for placement groups (programmatic interface only within the raylet).

**Comment by ericl** [comment]
@rkooo567 can you review this one?

File: src/ray/raylet/scheduling/cluster_resource_data.cc (hunk 1 above)
Comment (rkooo567) [reply src/ray/raylet/scheduling/cluster_resource_data.cc:87]: Will there be any issue if we make it return ID always?

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 3]
```diff
@@ -329,6 +329,34 @@ bool ClusterResourceScheduler::GetNodeResources(int64_t node_id,
 
 int64_t ClusterResourceScheduler::NumNodes() { return nodes_.size(); }
 
+void ClusterResourceScheduler::AddLocalResource(
+                                                      const std::string &resource_name,
+                                                      double resource_total) {
+  string_to_int_map_.Insert(resource_name);
+  int64_t resource_id = string_to_int_map_.Get(resource_name);
+
+  if (local_resources_.custom_resources.find(resource_id) != local_resources_.custom_resources.end()) {
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:338]:
```suggestion
  if (local_resources_.custom_resources.contains(resource_id)) {
```

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 4]
```diff
@@ -329,6 +329,34 @@ bool ClusterResourceScheduler::GetNodeResources(int64_t node_id,
...
+    instances.total[0] += total;
+    instances.available[0] += total;
+    auto &capacity = nodes_[local_node_id_].custom_resources[resource_id];
+    capacity.available += total;
+    capacity.total += total;
+  } else {
+    ResourceInstanceCapacities capacity;
+    capacity.total.resize(1);
+    capacity.total[0] = resource_total;
+    capacity.available.resize(1);
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:350]: This resize looks a little ugly. Any special reason why we don't just use emplace?

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 5]
```diff
@@ -329,6 +329,34 @@ bool ClusterResourceScheduler::GetNodeResources(int64_t node_id,
...
+    capacity.available[0] = resource_total;
+    local_resources_.custom_resources[resource_id] = capacity;
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:352]:
```suggestion
    local_resources_.custom_resources.emplace(resource_id, capacity);
```

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 6]
```diff
@@ -329,6 +329,34 @@ bool ClusterResourceScheduler::GetNodeResources(int64_t node_id,
...
+    std::string client_id_string = string_to_int_map_.Get(local_node_id_);
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:353]: This will have teh same issue that it won't return string if `local_node_id_ ` didn't exist right? Can you either add RAY_CHECK or change the behavior of this map to return string always?

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 7]
```diff
@@ -329,6 +329,34 @@ bool ClusterResourceScheduler::GetNodeResources(int64_t node_id,
...
+    RAY_CHECK(string_to_int_map_.Get(client_id_string) == local_node_id_);
+    UpdateResourceCapacity(string_to_int_map_.Get(local_node_id_), resource_name, resource_total);
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:355]:
```suggestion
    UpdateResourceCapacity(client_id_string, resource_name, resource_total);
```

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 8]
```diff
@@ -329,6 +329,34 @@ bool ClusterResourceScheduler::GetNodeResources(int64_t node_id,
...
+void ClusterResourceScheduler::AddLocalResource(
+                                                      const std::string &resource_name,
+                                                      double resource_total) {
+  string_to_int_map_.Insert(resource_name);
+  int64_t resource_id = string_to_int_map_.Get(resource_name);
+
+  if (local_resources_.custom_resources.find(resource_id) != local_resources_.custom_resources.end()) {
+    FixedPoint total(resource_total);
+    auto &instances = local_resources_.custom_resources[resource_id];
+    instances.total[0] += total;
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:341]: Can you comment what this index 0 mean?

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc (hunk 6 above)
Comment (rkooo567) [reply src/ray/raylet/scheduling/cluster_resource_scheduler.cc:353]:
```suggestion
    std::string node_id_string = string_to_int_map_.Get(local_node_id_);
```

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 9]
```diff
@@ -407,13 +441,23 @@ void ClusterResourceScheduler::DeleteResource(const std::string &node_id_string,
   };
   if (idx != -1) {
     it->second.predefined_resources[idx].total = 0;
+
+    if (node_id == local_node_id_) {
+      local_resources_.predefined_resources[idx].total.clear();
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:446]: Is it a bug you found? This code block makes sense, but it doesn't seem to be related to this PR.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.h [hunk 10]
```diff
@@ -201,6 +201,13 @@ class ClusterResourceScheduler {
   /// Get number of nodes in the cluster.
   int64_t NumNodes();
 
+  /// Add an (unused) local resource.
+  ///
+  /// \param resource_name: Resource which we want to update.
+  /// \param resource_total: New capacity of the resource.
+  void AddLocalResource(
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.h:208]: Why dont' we call it `AddcustomResources?`

File: src/ray/raylet/scheduling/cluster_resource_scheduler.h [hunk 11]
```diff
@@ -201,6 +201,13 @@ class ClusterResourceScheduler {
   /// Get number of nodes in the cluster.
   int64_t NumNodes();
 
+  /// Add an (unused) local resource.
```
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.h:204]: This description doesnt' sound clear to me. What does it mean by unused? Can you make it a little bit more clear?

**Review by rkooo567 — COMMENTED** [review]
Mostly clarification comment. Btw, isn't it more bug-free to have `UpdateLocalResources` than `AddLocalResources`, so that we can make the code more goal state-driven? Any thoughts about this? (Also, is there use case that we keep adding new custom resources)?

File: src/ray/raylet/scheduling/cluster_resource_data.cc (hunk 1 above)
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_resource_data.cc:87]: I think it's non trivial because StringIdMap takes an option argument `num_ids`

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc (hunk 8 above)
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_resource_scheduler.cc:341]: We always only use index 0 for non-unit resources (we do this in a lot of places)

```
src/ray/raylet/scheduling/cluster_task_manager_test.cc:459:            (expected_load[0] == load.num_infeasible_requests_queued() &&
src/ray/raylet/scheduling/cluster_task_manager_test.cc:467:        RAY_LOG(ERROR) << expected_load[0] << "\t" <<
src/ray/raylet/scheduling/cluster_resource_data.cc:72:    task_request.predefined_resources[0].demand = 0;
src/ray/raylet/scheduling/cluster_resource_data.cc:73:    task_request.predefined_resources[0].soft = false;
src/ray/raylet/scheduling/cluster_resource_scheduler.cc:443:    instance_list->total[0] = instance_list->available[0] = total;
src/ray/raylet/scheduling/cluster_resource_scheduler.cc:510:    if (available[0] >= remaining_demand) {
src/ray/raylet/scheduling/cluster_resource_scheduler.cc:511:      available[0] -= remaining_demand;
src/ray/raylet/scheduling/cluster_resource_scheduler.cc:512:      (*allocation)[0] = remaining_demand;
src/ray/raylet/scheduling/cluster_resource_scheduler.cc:516:        available[0] = 0;
```

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc (hunk 4 above)
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_resource_scheduler.cc:350]: I think resize handles more edge cases since we want this vector to always have length 1. It's a bit of a weird pattern the first time you see it, but this is also how we initialize resources in other parts of the code (see `InitResourceInstances`).

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc (hunk 9 above)
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_resource_scheduler.cc:446]: Yeah you can either think of it as a bug, or out of scope of what the function used to do. Either way, it's intention and necessary.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.h (hunk 10 above)
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_resource_scheduler.h:208]: Oh I think this would work for adding a predefined resource, but local is the important property because it avoids lookups and sets resource instances.

**Comment by wuisawesome** [comment]
> more goal state-driven

I think in general it's difficult to make accounting code like this goal state driven without introducing more race conditions. 

Having `AddLocalResources` **not** be idempotent is actually a feature here (in the context of placement groups). For example, we will call `AddLocalResources(wildcard_name, x);` multiple times for STRICT_PACK placement groups, and we _want_ to increase the available resources rather than just setting it.

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
LGTM. Seems like there are some conventions that are not good for readability, but anyway it is not in the scope of this PR.

**Comment by rkooo567** [comment]
There's a lint failure.
